File: haesa/0x05/2493.cpp
#include <bits/stdc++.h>
using namespace std;
stack<pair<int, int>> tower;
vector<int> result;
int n, h, max_k;
int pos;
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  cin >> n;
  tower.push({0, 100000001});
  for (int i = 1; i <= n; i++) {
    cin >> h;
    while (tower.top().second < h) tower.pop();
    cout << tower.top().first << ' ';
    tower.push({i, h});
  }
  cout << '\n';
}

File: haesa/0x05/2493.js
const input = require('fs')
  .readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split('\n');

const n = Number(input[0]);
const tower = input[1].split(' ').map(Number);
const stack = [];
const result = [];

for (let i = 0; i < n; i++) {
  while (stack.length && tower[stack[stack.length - 1]] < tower[i]) stack.pop();
  if (stack.length === 0) result.push(0);
  else result.push(stack[stack.length - 1] + 1);
  stack.push(i);
}
console.log(result.join(' '));
